crates/ide-completion/src/completions/attribute.rs

```diff
@@ -40,14 +41,23 @@ pub(crate) fn complete_known_attribute_input(
     extern_crate: Option<&ast::ExternCrate>,
 ) -> Option<()> {
     let attribute = fake_attribute_under_caret;
-    let name_ref = match attribute.path() {
-        Some(p) => Some(p.as_single_name_ref()?),
-        None => None,
-    };
-    let (path, tt) = name_ref.zip(attribute.token_tree())?;
+    let path = attribute.path()?;
+    let name_ref = path.segment()?.name_ref();
+    let (name_ref, tt) = name_ref.zip(attribute.token_tree())?;
```

Comment:
A nicer way to handle this I think would be something along the lines of 
```suggestion
    let path = attribute.path()?;
    let segments = path.segments().map(|s| s.name_ref()).collect::<Option<Vec<_>>>()?;
    let segments = segments.iter().map(|n| n.text().as_str()).collect::<ec<_>>();
    let tt = attribute.token_tree()?;
```

Then we can just match on `&*segments` (`&[&str]`), that is `["diagnostic", "on_unimplemented"] => ....`, etc which allows to easily combine the two `match`es here

Reply:
Ah right, `text()` may own